Feature/create qr code

jihwankim129 opened the pull request:
일단 Qr 코드 생성하는거 리뷰를 위해 드래프트 열어 놓겠습니다.

QrLink를 redis에서 가져오기 위한 key 값을 이제 membermeeting 안에 저장해야한다고 생각해서 membermeeting entity의 내용을 수정하였고, 

선택지는 두가지가 있는데 그냥 memberMeeting 내부의 필드에 string 타입으로 저장하거나 아니면

Qrlink 엔티티를 생성하여 1대1로 연결하고 QrLink엔티티 내부에 string 타입으로 키값을 저장하는 방법인데 

일단 후자 방법을 선택하였습니다. 그냥 memberMeeting내부 보니깐 최대한 연결고리라고 생각하고 다 엔티티 새로 만들어서 다 바깥으로 빼주길래, qr도 엔티티 만들어서 그냥 일대일로 연결시킨거고, 굳이 복잡하게이렇게 안하고 내부에 string 필드로 키값 멤버미팅에 저장해도 될것 같습니다.



키값을 클라이언트에 보내주면 클라이언트에서 키값으로 또 레디스에 접근해서 클라이언트 상에서 이미지를 가져오는 느낌으로 작성하였습니다.

pileuszu committed:
feat: build.gradle implementation 추가

- build.gradle: qr 에 필요한 implementation 추가

pileuszu committed:
feat: QrLink 생성, 조회, 검사 생성

- MemberMeeting.java: 필드값에 String QrLink 생성
- MemberMeetingRepository.java: findByMemberAndMeeting 쿼리문 생성
- QrController.java: QR, 생성, 조회, 검증 추가
- QrService.java: QrController.java에 필요한 함수 추가

pileuszu committed:
feat: redis 에 연결하기 위한 기본 파일 생성

- MemberMeeting.java: QrLink entity를 생성후 entity를 @One to One으로 연결
- MemberMeetingRepository.java: MemberMeetingRepository.java에 있는 내용을 QrLinkRedisRepository.java로 이동
- QrLink.java: 새로 Entity생성
- QrLinkRedisRepository.java: MemberMeetingRepository.java에 있는 내용을 QrLinkRedisRepository.java로 이동
- RedisConfig.java: Redis사용을 위한 Config 파일 추가

pileuszu committed:
feat: redis 연결

- MemberMeetingRepository.java: 다시 MemberMeetingRepository.java 의 query문 원상복귀
- value오류를 위한 io.swagger.v3.oas.annotations.responses.ApiResponses 추가
- QrLink.java: QrLink의 builder의 createQrLink문 추가
- QrLinkRedisRepository.java: query 삭제
- 오류로 인식되지않던 Qr코드 생성 오픈소스 import 추가

pileuszu committed:
fix: private 오타 수정

- QrService.java: private 오타 수정

jihwankim129 committed:
fix: QrService.java

- QrService.java: rediskey 생성에 QrLink id 가 아닌 memberMeeting id로 key 값을 생성하는 오타 수정

jihwankim129 commented:
아 근데 생각해보니 엔티티 자체에 redis 관련 어노테이션을 지정해야해서 이렇게 한것도 있네요

jihwankim129 committed:
fix: Qr 오류 수정

- QrController.java: reponses오타 수정
- RedisConfig.java: Value오타 수정

jihwankim129 committed:
fix: null 값을 참조하던 내용 수정

- MeetingService.java: findMeeting 과 validation 추가
- MemberMeetingRepository.java: 기존 쿼리문 삭제 및 상속받은 JPA 사용
- MemberMeetingService.java: MemberMeeting find 하는 것과 validation 추가
- MemberService.java: id 를 통해 찾을때 validation하는 내용 추가
- QrController.java: Contenttype image로 수정
- QrLink.java: QrLink Entity에 MemberMeeting set하는 내용 추가
- QrService.java: 필요한 Service단 추가 및 수정

# Conflicts:
#	back-end/src/main/java/kr/co/ssalon/domain/repository/MemberMeetingRepository.java

jihwankim129 commented:
qr 에 대해 생성 및 검증하는 것은 잘 되는것을 확인하였고 다만 풀 리퀘스트 승인 시점을 멤버와 미팅에 대한 추가 삭제 가 가능하여 실제로 값을 넣어 볼수 있을 시점 이후로 하면 좋을 것같습니다.

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/entity/QrLink.java:
```diff
@@ -0,0 +1,39 @@
+    @OneToOne(fetch = FetchType.LAZY)
+    private MemberMeeting memberMeeting;
+
+    private String QrLink;
```
필드명은 소문자로 시작해야 맞는 거 같아요.

jihwankim129 replied:
아하 소문자로 변경하겠습니다.

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/entity/QrLink.java:
```diff
@@ -0,0 +1,39 @@
+        QrLink qrLinkEntity = new QrLink();
+        qrLinkEntity.setMemberMeeting(memberMeeting);
+        qrLinkEntity.setQrLink(qrLink);
+        return qrLinkEntity;
```
QrLink 필드명을 소문자로 시작하도록 수정 후에, QrLink 엔티티는 다른 엔티티와 통일성을 위해 builder 사용해서 생성하면 좋겠네요.

jihwankim129 replied:
넵!

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/MeetingService.java:
```diff
@@ -53,4 +55,17 @@ public Page<Meeting> getMoims(MeetingSearchCondition meetingSearchCondition, Pag
+        return meeting;
+    }
+
+    private Meeting validaitonMeeting(Optional<Meeting> meeting) throws BadRequestException {
```
오타났네요. `MemberService.validaitonMeeting`에도 오타가 같이 났었군요. 같이 수정해주시면 감사하겠습니다!
```suggestion
    private Meeting validationMeeting(Optional<Meeting> meeting) throws BadRequestException {
```

jihwankim129 replied:
헉 수정하겠습니다

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/MemberMeetingService.java:
```diff
@@ -0,0 +1,28 @@
+        if (findMemberMeeting.isPresent()) {
+            return findMemberMeeting.get();
+        } else
+            throw new BadRequestException("멤버 미팅을 찾을 수 없습니다.");
```
클라이언트에서는 '멤버 미팅'이 무엇을 뜻하는지 모를 수 있어서, 아래처럼 좀더 일반적인 문구로 써주면 좋을 것 같아요.
```suggestion
            throw new BadRequestException("회원이 참여하고 있는 모임을 찾을 수 없습니다.");
```

jihwankim129 replied:
아하 넵 변경하겠습니다

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/QrService.java:
```diff
@@ -0,0 +1,122 @@
+    private MemberService memberService;
+
+    @Autowired
+    private MeetingService meetingService;
```
Autowired를 통한 의존성 주입은 여러가지 측면에서 지양해야 하는 것으로 알고 있습니다. Autowired 주입과 생성자 주입의 차이점을 한 번 보시는 것도 좋겠네요. 저희 프로젝트의 다른 Service를 보시면 S3 부분을 제외하고 생성자 주입을 사용하고 있습니다. 혹시 이 QR 관련된 부분은 Autowired를 사용해야만 하는 이유가 있는 걸까요?
```suggestion
@RequiredArgsConstructor
public class QrService {

    private final RedisTemplate<String, byte[]> redisTemplate;
    private final MeetingRepository meetingRepository;
    private final MemberRepository memberRepository;
    private final MemberMeetingRepository memberMeetingRepository;
    private final MemberService memberService;
    private final MeetingService meetingService;
```

jihwankim129 replied:
아래 수정된 버전으로 바꿀게요

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/QrService.java:
```diff
@@ -0,0 +1,122 @@
+    private MeetingService meetingService;
+
+
+    private MemberMeetingService memberMeetingService;
```
이 부분은 Autowired가 붙지 않았는데, 주입받지 않아도 되는 부분일까요?

lee1684 replied:
@jihwankim129 여기는 왜 안 보이는지 모르겠지만 이 리뷰도 봐주세요~

jihwankim129 replied:
아하 HIde 도 열어서 한번씩 확인하고 있어서 확인했습니다! 읽어볼게요!

jihwankim129 replied:
일단 위에 처럼 Auto wired 없이 private final로 해놓을게요

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/QrService.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.concurrent.TimeUnit;
+
+@Service
+@RestController
```
Service 파일인데 Controller 관련 어노테이션이 붙어 있네요?

jihwankim129 replied:
앗 저게 왜붙어잇지

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/QrService.java:
```diff
@@ -0,0 +1,122 @@
+        MemberMeeting memberMeeting = memberMeetingService.findByMemberAndMeeting(member, meeting);
+
+        try {
+            return redisTemplate.opsForValue().get(memberMeeting.getQrLink().getQrLink());
```
getQrLink()가 두 번 들어가서 처음에 오타인가 했는데 QrLink 내부에 qrLink 필드가 있어서 이렇게 됐군요.
QrLink의 qrLink 필드명은 가독성을 위해 수정이 필요할 듯 싶은데 필드명을 link로 하는 게 어떨까요?

jihwankim129 replied:
가독성이 어려운것 같긴하네요 아예 키값을 저장한다는 뜻으로  QrKey 이런식으로 할까요?

lee1684 replied:
넵

lee1684 commented on back-end/src/main/java/kr/co/ssalon/config/RedisConfig.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Bean
+    public RedisTemplate<?, ?> redisTemplate() {
+        RedisTemplate<?, ?> redisTemplate = new RedisTemplate<>();
```
`RedisTemplate<String, byte[]>` 이렇게 쓰이는 부분 말고는 없는 것 같은데 타입을 `?`로 해야 하는 이유가 혹시 따로 있나요?

jihwankim129 replied:
? 로 놓으면 나중에 재사용도 가능해질것같아서 이렇게 했었어요

lee1684 replied:
그렇군요

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/QrService.java:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    @Transactional
+    public boolean checkQrLink(@AuthenticationPrincipal CustomOAuth2Member customOAuth2Member, Long moimId, Long userId, MultipartFile file) throws BadRequestException {
```
이 메소드에만 Transactional이 붙어 있는데 이유가 있을까요?

jihwankim129 replied:
다른 메소드에도 transactional을 붙일게요

jihwankim129 replied:
아 근데 그냥 여기선 굳이 전체적으로 transactional이 필요가 없나

jihwankim129 replied:
일단 굳이 안붙이고 다 지워놓는걸로 그냥 할게요

lee1684 replied:
이 메소드 안에서 여러 번의 DB 접근을 하는 것 같은데, Transactional 붙이는 게 좋을 것 같지 않나요??

> 비정상적 종료로 인한 rollback이 발생할 경우에는 트랜잭션의 일부 작업만 데이터베이스에 반영되는 것을 방지해 데이터 일관성을 유지할 수 있습니다.

jihwankim129 replied:
그래서 원래는 저 왠만해선 이유없어도 transactional 메소드에 다 붙여놓을려고 하는데 어떤가요.

다른 메소드에 없던거는 실수로 빠졌네요

lee1684 replied:
넵

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/QrService.java:
```diff
@@ -0,0 +1,122 @@
+        } catch (Exception e) {
+            // QR 코드 생성에 실패한 경우 예외 처리
+            e.printStackTrace();
+            return new byte[0];
```
QR 코드 생성에 실패한 경우에도 예외 throw가 아니라 객체 생성을 하고 있네요?
그러면 redis에 빈 byte가 저장되는데 의도하신 부분일까요?

jihwankim129 replied:
음 throw를 던지는 걸루 수정하는게 좋을것같아요

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/repository/QrLinkRedisRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface QrLinkRedisRepository extends CrudRepository<QrLink, String> {
+
+}
```
이 QrLinkRedisRepository은 사용이 아예 안 되고 있는데 생성됐네요?

jihwankim129 replied:
음 저거는 일단 지워볼게요

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/entity/QrLink.java:
```diff
@@ -0,0 +1,39 @@
+@Builder
+@AllArgsConstructor
+@RedisHash(value = "QrLink")
+public class QrLink {
```
이 @RedisHash 어노테이션은 QrLink 객체 그 자체를 Redis에 저장하기 위한 어노테이션으로 알고 있는데, 현재 작성하신 방향은 아래 코드처럼 QR의 byte만을 Redis에 저장하는 것 같아요. 그럼 이 부분은 제거해도 될까요?
```
            // QR 이미지 생성
            byte[] qrImage = generateQRCode(randomStr);

            // Redis에 QR 이미지 저장
            String redisKey = "QR_" + memberMeeting.getId();
            redisTemplate.opsForValue().set(redisKey, qrImage);
```

jihwankim129 replied:
아하 제가 저 redis hash를 제대로 이해를 못했어서, 그럼 없어도 되는거군요. 저는 접근하기 위해 붙여야하는줄 알았어요

lee1684 replied:
저도 정확하지 않을 수 있어서 테스트 해봐야겠네요.

lee1684 replied:
이 부분은 테스트 아직 안된 거죠?

lee1684 replied:
일단 이렇다네요.
![image](https://github.com/lee1684/SKYTeam/assets/162105892/304bdbaf-9184-4ec1-8bec-7cdca1a002ad)

jihwankim129 replied:
ㅇㅎㅇㅎ 그냥 저거 지울게영

lee1684 replied:
넵

pileuszu committed:
fix:

- MeetingService.java: validation 오타 수정
- MemberMeetingService.java: validationMemberMeeting BadRequestException내용 변경 "멤버 미팅을 찾을 수 없습니다." -> "회원이 참여하고 있는 모임을 찾을 수 없습니다."
- MemberService.java: validation 오타 수정
- QrLink.java: createQrLink에서 settter 를 사용하지 않고 build사용
- QrService.java: Autowired annotation을 지우고 private final로 수정, e.printStackTrac()를 지우고 예외를 던지는 걸로 수정

pileuszu committed:
fix: pr 내용 반영

- MeetingService.java: validation 오타 수정
- MemberMeetingService.java: validationMemberMeeting BadRequestException내용 변경 "멤버 미팅을 찾을 수 없습니다." -> "회원이 참여하고 있는 모임을 찾을 수 없습니다."
- MemberService.java: validation 오타 수정
- QrLink.java: createQrLink에서 settter 를 사용하지 않고 build사용
- QrService.java: Autowired annotation을 지우고 private final로 수정, e.printStackTrac()를 지우고 예외를 던지는 걸로 수정, @RequiredArgsConstructor 어노테이션 추가, getQrLink()를 getQrKey()로 수정
- QrLinkRedisRepository.java: 삭제

jihwankim129 committed:
fix: pr 내용 반영 redhash 제거

- QrLink.java: redhash 제거

jihwankim129 committed:
fix: pr 반영 후 기타 오류 해결

- MemberMeetingService.java: MemberMeetingService annotation설정

lee1684 commented on back-end/src/main/java/kr/co/ssalon/domain/service/MemberMeetingService.java:
```diff
@@ -4,13 +4,19 @@
 
+@Service
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
```
현재 MemberMeetingService에는 조회하는 메소드만 있어서 readOnly 붙이신 걸까요?
추후에 조회 이외의 메소드들도 추가될 가능성은 없을까요?

lee1684 replied:
readOnly 옵션 제거해서 [새로운 PR](https://github.com/lee1684/SKYTeam/pull/18)에 반영해두었습니다.
